### PR TITLE
S3 batch

### PR DESCRIPTION
**Pull request by msailes**
*Issue #, if available:*

https://github.com/aws/aws-lambda-java-libs/issues/177

*Description of changes:*

Event and response objects for S3 Batch operations.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/S3BatchResponse.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+@Data
+@Builder(setterPrefix = "with")
+@NoArgsConstructor
+@AllArgsConstructor
+public class S3BatchResponse {
+
+    private String invocationSchemaVersion;
+    private String treatMissingKeysAs;
+    private String invocationId;
+    private List<Result> result;
```
Comment (david-kingsbury-bah) [inline aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/S3BatchResponse.java:26]:
```suggestion
    private List<Result> results;
```

Comment (david-kingsbury-bah) [reply aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/S3BatchResponse.java:26]: Plural "results"

**Review by david-kingsbury-bah — CHANGES_REQUESTED** [review]
After making the one field name change (result VS results), I was able to successfully run an S3 Batch Operations job that invokes a Lambda written using these models.

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/S3BatchResponse.java (hunk 1 above)
Comment (msailes) [reply aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/S3BatchResponse.java]: fixed.

**Comment by david-kingsbury-bah** [comment]
Since we have the ResultCode enum now, I'm thinking S3BatchOpResponse.treatMissingKeysAs should be of type ResultCode.

```
private ResultCode treatMissingKeysAs;
```

### The following are ideas that might go too far for this library; let me know what you think.

We could default treatMissingKeysAs to PermanentFailure, as this is _likely_ the value _most_ people will use.

```
private ResultCode treatMissingKeysAs = ResultCode.PermanentFailure;
```

We could have a special constructor to seed a response instance. These must be set for the response to be considered valid. If one does not want to use this, they can use the other constructors.

```
    public S3BatchResponse(final S3BatchEvent request) {
        this.invocationSchemaVersion = request.getInvocationSchemaVersion();
        this.invocationId = request.getInvocationId();
        this.results = Collections.singletonList(
                Result.builder()
                        .withTaskId(request.getTasks().get(0).getTaskId())
                        .withResultCode(ResultCode.Succeeded)
                        .build()
        );
    }
```

**Comment by msailes** [comment]
Hi David,

Great feedback! What about something like this in the S3BatchResponse.java class.

```java
public static S3BatchResponseBuilder fromS3BatchEvent(S3BatchEvent s3BatchEvent) {
        return S3BatchResponse.builder()
                .withInvocationId(s3BatchEvent.getInvocationId())
                .withInvocationSchemaVersion(s3BatchEvent.getInvocationSchemaVersion());
}
```

I think I'd be nervous adding any default behavior which might have a negative impact. I don't know this area of the platform that well, so I would definitely appreciate any feedback.

Thanks,

Mark

**Comment by david-kingsbury-bah** [comment]
The `fromS3BatchEvent` method looks great!

I think the question now is should `fromS3BatchEvent` go as far as my example and create a single `Result` with the Task ID set.

- S3 Batch Ops expects the response result Task ID to match what was passed in through the request (`S3BatchEvent`). I suspect this allows S3 Batch Ops to do many concurrent invocations and still be able to gather and sort results.
- S3 Batch Ops only sends 1 task, and expects only 1 result. Here is a snippet from their docs
   ```
   # S3 batch operations currently only passes a single task at a time in the array of tasks.
   task = event['tasks'][0]
   ```

I'm on the fence. The API _allows_ for multiple tasks and results, however it is certainly 1 task per event today.

**Comment by msailes** [comment]
In my opinion no. 

If all that was required was the task id it would make sense. But the user will still have to add their own values for resultCode and resultString. So I don't think there is much to be gained.

**Comment by msailes** [comment]
I've also tested this now.

**Review by david-kingsbury-bah — APPROVED** [review]

**Review by carlzogh — APPROVED** [review]
LGTM!

**Comment by carlzogh** [comment]
Thanks @msailes and @david-kingsbury-bah! :)

**Comment by david-kingsbury-bah** [comment]
Excellent!

Resolved #177

**Pull request merged**